elasticdl/python/tests/odps_data_reader_test.py

```diff
@@ -0,0 +1,111 @@
+            }
+            example = tf.train.Example(
+                features=tf.train.Features(feature=example_dict)
+            )
+            f.write(example.SerializeToString())
+    return temp_file.name
+
+
+class DataReaderTest(unittest.TestCase):
+    Task = namedtuple("Task", ["start", "end", "shard_name"])
+
+    def test_recordio_data_reader(self):
+        num_records = 128
+        with tempfile.TemporaryDirectory() as temp_dir_name:
+            shard_name = _create_recordio_file(num_records, temp_dir_name)
+
+            # Test shards creation
+            expected_shards = {shard_name: (0, num_records)}
+            reader = RecordIODataReader(data_dir=temp_dir_name)
+            self.assertEqual(expected_shards, reader.create_shards())
+
+            # Test records reading
+            records = list(
+                reader.read_records(self.Task(0, num_records, shard_name))
+            )
+            self.assertEqual(len(records), num_records)
+            for record in records:
+                parsed_record = tf.io.parse_single_example(
+                    record,
+                    {
+                        "x": tf.io.FixedLenFeature([1], tf.float32),
+                        "y": tf.io.FixedLenFeature([1], tf.float32),
+                    },
+                )
+                for k, v in parsed_record.items():
+                    self.assertEqual(len(v.numpy()), 1)
+
+    @unittest.skipIf(
+        os.environ.get("ODPS_TESTS", "False") == "False",
+        "ODPS environment is not configured",
+    )
+    def test_odps_data_reader(self):
+        # Setup
+        project = os.environ[ODPSConfig.PROJECT_NAME]
+        access_id = os.environ[ODPSConfig.ACCESS_ID]
+        access_key = os.environ[ODPSConfig.ACCESS_KEY]
+        endpoint = os.environ.get(ODPSConfig.ENDPOINT)
+        test_table = "test_odps_data_reader_%d_%d" % (
+            int(time.time()),
+            random.randint(1, 101),
+        )
+        odps_client = ODPS(access_id, access_key, project, endpoint)
+        create_iris_odps_table(odps_client, project, test_table)
+        records_per_task = 50
+
+        reader = ODPSDataReader(
+            project=project,
+            access_id=access_id,
+            access_key=access_key,
+            endpoint=endpoint,
+            table=test_table,
+            num_processes=1,
+            records_per_task=records_per_task,
+        )
+
+        # Test shards creation
+        expected_shards = {
+            "shard_0": (0, records_per_task),
+            "shard_1": (50, records_per_task),
+            "shard_2": (100, 10),
+        }
+        self.assertEqual(expected_shards, reader.create_shards())
+
+        # Test records reading
+        records = list(reader.read_records(self.Task(0, 2, "shard_0")))
+        self.assertEqual(
+            [[6.4, 2.8, 5.6, 2.2, 2], [5.0, 2.3, 3.3, 1.0, 1]], records
+        )
+
+        # Teardown
```

Comment:
Is it better we use `tearDown` from `TestCase`? I am afraid if any test fails, cleaning up operations will not be executed.

Reply:
Done. Thanks.